File: java/MoreThanHalf/MoreThanHalf.py
def partition(li, start, end):
    value = li[start]
    while start < end:
        while li[end] >= value and start < end:
            end -= 1
        li[start] = li[end]
        while li[start] <= value and start < end:
            start += 1
        li[end] = li[start]
    li[start] = value
    return start


def sort(li, start, end):
    if start >= end:
        return
    middle = partition(li, start, end)
    sort(li, start, middle - 1)
    sort(li, middle + 1, end)


def verify(li, num):
    i = 0
    for value in li:
        if value == num:
            i += 1
    if i > len(li)/2:
        return True
    else:
        return False


def MoreThanHalf(li):
    middle = len(li) // 2
    start = 0
    end = len(li) - 1
    index = partition(li, start, end)
    while index != middle:
        if index > middle:
            end = index - 1
            index = partition(li, start, end)
        else:
            start = index + 1
            index = partition(li, start, end)
    num = li[index]
    if verify(li, num):
        return num
    else:
        return 0


print(MoreThanHalf([1, 2, 3, 2, 2, 2, 5, 4, 2]))

File: java/MoreThanHalf/MoreThanHalf.java
package question29;

/**
 * Created by Claudius on 2017/5/3.
 */
public class MoreThanHalf {
    public static void main(String[] args) {
        int[] nums = {1, 2, 3, 2, 2, 2, 5, 4, 2};
        System.out.print(caculator(nums));
    }

    public static int caculator(int[] nums) {
        int middle = nums.length >> 1;
        int start = 0;
        int end = nums.length - 1;
        int index = partition(nums, start, end);
        while (index != middle) {
            if (index > middle) {
                end = index - 1;
                index = partition(nums, start, end);
            } else {
                start = index + 1;
                index = partition(nums, start, end);
            }
        }
        int result = nums[index];
        if (verify(nums, result))
            return result;
        else
            return 0;
    }

    public static boolean verify(int nums[], int number) {
        int time = 0;
        int length = nums.length;
        for (int i = 0; i < length; i++) {
            if (nums[i] == number)
                time++;
        }
        if (time > length >> 1)
            return true;
        else
            return false;
    }

    public static int partition(int[] nums, int start, int end) {
        int value = nums[start];
        int left = start, right = end;
        while (left < right) {
            while (left < right && nums[right] >= value)
                right--;
            nums[left++] = nums[right];
            while (left < right && nums[left] <= value)
                left++;
            nums[right--] = nums[left];
        }
        nums[left] = value;
        return left;
    }
}
